File: apps/rom-manager-ui/server/remote-zip.js
// Remote ZIP reader — list and extract files from a ZIP archive over HTTP
// without downloading the whole file. Uses HTTP Range requests.

export class RemoteZip {
  constructor(url) {
    this.url = url;
    this._size = null;
    this._entries = null;
  }

  async _fetchRange(start, end) {
    let resp = await fetch(this.url, {
      headers: { Range: `bytes=${start}-${end}` },
      redirect: 'manual',
    });
    // Handle IA's redirect to CDN
    if (resp.status >= 300 && resp.status < 400) {
      const location = resp.headers.get('location');
      if (!location) throw new Error('Redirect without location');
      this.url = location;
      resp = await fetch(this.url, {
        headers: { Range: `bytes=${start}-${end}` },
      });
    }
    if (!resp.ok && resp.status !== 206) throw new Error(`HTTP ${resp.status} requesting range ${start}-${end}`);
    return new Uint8Array(await resp.arrayBuffer());
  }

  async size() {
    if (this._size) return this._size;
    // GET first byte to detect file size (more reliable than HEAD for IA redirects)
    const resp = await fetch(this.url, {
      headers: { Range: 'bytes=0-0' },
    });
    if (resp.status !== 206 && !resp.ok) throw new Error(`HTTP ${resp.status} for ${this.url}`);
    const s = parseInt(resp.headers.get('content-range')?.split('/')[1] || resp.headers.get('content-length'), 10);
    if (!s) throw new Error('Could not determine file size');
    this._size = s;
    return s;
  }

  async readU32(arr, off) { return (arr[off] | (arr[off + 1] << 8) | (arr[off + 2] << 16) | (arr[off + 3] << 24)) >>> 0; }
  async readU16(arr, off) { return (arr[off] | (arr[off + 1] << 8)) >>> 0; }
  async readU64(arr, off) {
    const low = (arr[off] | (arr[off + 1] << 8) | (arr[off + 2] << 16) | (arr[off + 3] << 24)) >>> 0;
    const high = (arr[off + 4] | (arr[off + 5] << 8) | (arr[off + 6] << 16) | (arr[off + 7] << 24)) >>> 0;
    return low + high * 0x100000000;
  }

  async _loadCentralDir() {
    const totalSize = await this.size();

    // Read last 128KB to find EOCD and ZIP64 locator
    const tailLen = Math.min(totalSize, 131072);
    const tail = await this._fetchRange(totalSize - tailLen, totalSize - 1);
    const tailOff = totalSize - tailLen;

    let cdOffset, cdSize, cdCount;

    // Try ZIP64 first: find PK\x06\x07 (ZIP64 End of Central Directory Locator)
    let zip64LocOff = -1;
    for (let i = tail.length - 20; i >= 0; i--) {
      if (tail[i] === 0x50 && tail[i + 1] === 0x4b && tail[i + 2] === 0x06 && tail[i + 3] === 0x07) {
        zip64LocOff = tailOff + i;
        break;
      }
    }
    if (zip64LocOff >= 0) {
      const z64local = zip64LocOff - tailOff;
      const z64EocdOff = await this.readU64(tail, z64local + 8);
      // Download ZIP64 EOCD (fixed 56 bytes + variable size)
      const z64Eocd = await this._fetchRange(z64EocdOff, z64EocdOff + 63);
      const z64EocdSize = await this.readU64(z64Eocd, 4); // size of ZIP64 EOCD
      cdCount = await this.readU64(z64Eocd, 24); // total entries
      cdSize = await this.readU64(z64Eocd, 40); // central directory size
      cdOffset = await this.readU64(z64Eocd, 48); // central directory offset
    } else {
      // Fallback to standard EOCD
      let eocdOff = -1;
      for (let i = tail.length - 22; i >= 0; i--) {
        if (tail[i] === 0x50 && tail[i + 1] === 0x4b && tail[i + 2] === 0x05 && tail[i + 3] === 0x06) {
          eocdOff = tailOff + i;
          break;
        }
      }
      if (eocdOff === -1) throw new Error('EOCD not found');
      const eocdLocalOff = eocdOff - tailOff;
      cdOffset = await this.readU32(tail, eocdLocalOff + 16);
      cdSize = await this.readU32(tail, eocdLocalOff + 12);
      cdCount = await this.readU16(tail, eocdLocalOff + 8);
    }

    // Download central directory
    const cd = await this._fetchRange(cdOffset, cdOffset + cdSize - 1);

    // Parse entries
    const entries = [];
    let pos = 0;
    for (let i = 0; i < cdCount; i++) {
      if (cd[pos] !== 0x50 || cd[pos + 1] !== 0x4b || cd[pos + 2] !== 0x01 || cd[pos + 3] !== 0x02) break;
      const versionMadeBy = await this.readU16(cd, pos + 4);
      const versionNeeded = await this.readU16(cd, pos + 6);
      const flags = await this.readU16(cd, pos + 8);
      const compMethod = await this.readU16(cd, pos + 10);
      const lastMod = await this.readU16(cd, pos + 12);
      const crc32 = await this.readU32(cd, pos + 16);
      const compSize = await this.readU32(cd, pos + 20);
      const uncompSize = await this.readU32(cd, pos + 24);
      const nameLen = await this.readU16(cd, pos + 28);
      const extraLen = await this.readU16(cd, pos + 30);
      const commentLen = await this.readU16(cd, pos + 32);
      const localOff = await this.readU32(cd, pos + 42);
      const nameBytes = cd.slice(pos + 46, pos + 46 + nameLen);
      const name = new TextDecoder().decode(nameBytes);

      // Check ZIP64 extra field for large values
      let actualLocalOff = localOff;
      let actualCompSize = compSize;
      let actualUncompSize = uncompSize;
      if (localOff === 0xFFFFFFFF || compSize === 0xFFFFFFFF || uncompSize === 0xFFFFFFFF) {
        const extra = cd.slice(pos + 46 + nameLen, pos + 46 + nameLen + extraLen);
        let epos = 0;
        while (epos < extra.length - 3) {
          const eid = await this.readU16(extra, epos);
          const esize = await this.readU16(extra, epos + 2);
          if (eid === 0x0001) { // ZIP64 extended information
            let dp = epos + 4;
            if (actualUncompSize === 0xFFFFFFFF) { actualUncompSize = await this.readU64(extra, dp); dp += 8; }
            if (actualCompSize === 0xFFFFFFFF) { actualCompSize = await this.readU64(extra, dp); dp += 8; }
            if (actualLocalOff === 0xFFFFFFFF) { actualLocalOff = await this.readU64(extra, dp); dp += 8; }
            break;
          }
          epos += 4 + esize;
        }
      }

      entries.push({
        name,
        localOffset: actualLocalOff,
        compressedSize: actualCompSize,
        uncompressedSize: actualUncompSize,
        method: compMethod,
      });
      pos += 46 + nameLen + extraLen + commentLen;
    }
    this._entries = entries;
    return entries;
  }

  async listFiles(pattern) {
    const entries = this._entries || (await this._loadCentralDir());
    if (!pattern) return entries;
    const re = pattern instanceof RegExp ? pattern : new RegExp(pattern.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i');
    return entries.filter(e => re.test(e.name));
  }

  async extractToFile(entryName, destPath) {
    const entries = this._entries || (await this._loadCentralDir());
    const entry = entries.find(e => e.name === entryName);
    if (!entry) throw new Error(`Entry "${entryName}" not found in remote zip`);

    // Read local file header + data
    const headerLen = 30; // minimum local header size
    const nameLen = entry.name.length;
    const localHeader = await this._fetchRange(entry.localOffset, entry.localOffset + headerLen + nameLen - 1);
    const extraLen = await this.readU16(localHeader, 28);
    const dataStart = entry.localOffset + headerLen + nameLen + extraLen;

    // Download data
    const data = await this._fetchRange(dataStart, dataStart + entry.compressedSize - 1);

    // Decompress if needed
    let finalData = data;
    if (entry.method === 8) {
      const zlib = await import('zlib');
      finalData = await new Promise((resolve, reject) => {
        zlib.inflateRaw(data, (err, result) => {
          if (err) reject(err);
          else resolve(result);
        });
      });
    } else if (entry.method !== 0) {
      throw new Error(`Unsupported compression method: ${entry.method}`);
    }

    const fs = await import('fs');
    const path = await import('path');
    fs.mkdirSync(path.dirname(destPath), { recursive: true });
    fs.writeFileSync(destPath, Buffer.from(finalData));
    return { name: entryName, size: finalData.length };
  }
}

File: libs/ia-archive/src/remote_zip.rs
/// Minimal ZIP entry info parsed from central directory
#[derive(Debug, Clone)]
pub struct ZipEntry {
    pub name: String,
    pub crc32: Option<u32>,
    pub compressed_size: u64,
    pub uncompressed_size: u64,
    pub local_offset: u64,
}

/// Reads a remote ZIP's central directory via HTTP Range requests.
pub struct RemoteZip {
    url: String,
    client: reqwest::Client,
}

impl RemoteZip {
    pub fn new(url: &str) -> Self {
        Self {
            url: url.to_string(),
            client: reqwest::Client::new(),
        }
    }

    /// Create a RemoteZip with an authenticated client (cookie jar included)
    pub fn new_with_client(url: &str, client: reqwest::Client) -> Self {
        Self {
            url: url.to_string(),
            client,
        }
    }

    /// Fetch a byte range from the remote file, following IA redirects
    async fn fetch_range(&mut self, start: u64, end: u64) -> Result<Vec<u8>, String> {
        for _ in 0..5 {
            let resp = self
                .client
                .get(&self.url)
                .header("Range", format!("bytes={}-{}", start, end))
                .header("User-Agent", "GameManager/0.1")
                .send()
                .await
                .map_err(|e| format!("HTTP error: {}", e))?;

            if resp.status().is_redirection() {
                if let Some(location) = resp.headers().get("location") {
                    if let Ok(loc) = location.to_str() {
                        self.url = loc.to_string();
                        continue;
                    }
                }
            }
            if !resp.status().is_success() && resp.status().as_u16() != 206 {
                return Err(format!("Range request returned HTTP {}", resp.status()));
            }
            return resp
                .bytes()
                .await
                .map(|b| b.to_vec())
                .map_err(|e| format!("Read error: {}", e));
        }
        Err("Too many redirects".into())
    }

    async fn get_size(&mut self) -> Result<u64, String> {
        for _ in 0..5 {
            let resp = self
                .client
                .get(&self.url)
                .header("Range", "bytes=0-0")
                .header("User-Agent", "GameManager/0.1")
                .send()
                .await
                .map_err(|e| format!("HTTP error: {}", e))?;

            if resp.status().is_redirection() {
                if let Some(location) = resp.headers().get("location") {
                    if let Ok(loc) = location.to_str() {
                        self.url = loc.to_string();
                        continue;
                    }
                }
            }
            if let Some(val) = resp.headers().get("content-range") {
                let s = val.to_str().map_err(|_| "Bad content-range header")?;
                let total = s.rsplit('/').next().unwrap_or("0");
                return total.parse::<u64>().map_err(|_| "Bad content-range value".into());
            }
            return Err("No content-range header".into());
        }
        Err("Too many redirects".into())
    }

    /// Parse the central directory and return all zip entries.
    pub async fn list_entries(&mut self) -> Result<Vec<ZipEntry>, String> {
        let size = self.get_size().await?;
        let read_size = std::cmp::min(size, 131072);
        let tail = self.fetch_range(size - read_size, size - 1).await?;

        let eocd_offset = tail
            .windows(4)
            .enumerate()
            .rev()
            .find(|(_, w)| *w == [0x50, 0x4b, 0x05, 0x06])
            .map(|(pos, _)| pos);

        let (cd_offset, _num_entries) = match eocd_offset {
            Some(pos) => {
                let cd_offset = u64::from_le_bytes([
                    tail[pos + 20], tail[pos + 21], tail[pos + 22], tail[pos + 23],
                    0, 0, 0, 0,
                ]);
                let num = u64::from_le_bytes([
                    tail[pos + 12], tail[pos + 13], tail[pos + 14], tail[pos + 15],
                    0, 0, 0, 0,
                ]);
                (cd_offset, num)
            }
            None => return Err("Could not find EOCD record".into()),
        };

        // Read central directory size from EOCD
        let cd_size_pos = eocd_offset.unwrap();
        let cd_size = u64::from_le_bytes([
            tail[cd_size_pos + 16], tail[cd_size_pos + 17],
            tail[cd_size_pos + 18], tail[cd_size_pos + 19],
            0, 0, 0, 0,
        ]);

        let cd_data = self.fetch_range(cd_offset, cd_offset + cd_size - 1).await?;

        let mut entries = Vec::new();
        let mut pos = 0usize;

        while pos + 46 <= cd_data.len() {
            let sig = &cd_data[pos..pos + 4];
            if sig != &[0x50, 0x4b, 0x01, 0x02] {
                break;
            }

            let crc = u32::from_le_bytes([
                cd_data[pos + 16], cd_data[pos + 17],
                cd_data[pos + 18], cd_data[pos + 19],
            ]);
            let comp_size = u32::from_le_bytes([
                cd_data[pos + 20], cd_data[pos + 21],
                cd_data[pos + 22], cd_data[pos + 23],
            ]);
            let uncomp_size = u32::from_le_bytes([
                cd_data[pos + 24], cd_data[pos + 25],
                cd_data[pos + 26], cd_data[pos + 27],
            ]);
            let name_len = u16::from_le_bytes([cd_data[pos + 28], cd_data[pos + 29]]) as usize;
            let extra_len = u16::from_le_bytes([cd_data[pos + 30], cd_data[pos + 31]]) as usize;
            let _comment_len = u16::from_le_bytes([cd_data[pos + 32], cd_data[pos + 33]]) as usize;
            let local_offset = u32::from_le_bytes([
                cd_data[pos + 42], cd_data[pos + 43],
                cd_data[pos + 44], cd_data[pos + 45],
            ]);

            if name_len > 0 && pos + 46 + name_len <= cd_data.len() {
                let name_bytes = &cd_data[pos + 46..pos + 46 + name_len];
                let name = String::from_utf8_lossy(name_bytes).to_string();
                entries.push(ZipEntry {
                    name,
                    crc32: Some(crc),
                    compressed_size: comp_size as u64,
                    uncompressed_size: uncomp_size as u64,
                    local_offset: local_offset as u64,
                });
            }

            pos += 46 + name_len + extra_len + _comment_len;
        }

        Ok(entries)
    }

    pub fn find_entry<'a>(&self, entries: &'a [ZipEntry], name: &str) -> Option<&'a ZipEntry> {
        entries.iter().find(|e| {
            e.name == name || e.name.ends_with(&format!("/{}", name))
                || e.name.ends_with(name)
        })
    }

    pub async fn extract_entry(&mut self, entry: &ZipEntry) -> Result<Vec<u8>, String> {
        let header_size = 30u64;
        let header = self.fetch_range(entry.local_offset, entry.local_offset + header_size + 200).await?;
        let name_len = u16::from_le_bytes([header[26], header[27]]) as u64;
        let extra_len = u16::from_le_bytes([header[28], header[29]]) as u64;
        let data_start = entry.local_offset + 30 + name_len + extra_len;
        let data_end = data_start + entry.compressed_size - 1;
        let data = self.fetch_range(data_start, data_end).await?;
        Ok(data)
    }
}
